### ISSUE TITLE
Area2D doesn't detect static bodies if "Monitorable" is disabled

### ISSUE THREAD

**Issue by YeldhamDev** [godotengine/godot#17238]
**Godot version:**
 184b2fe

**Issue description:**
This could also affect the 3D version of it, but I just tested in 2D.

If an Area2D's `Monitorable` property is disabled, it will fail to detect static bodies.

**Comment by Ranoller** [comment]
http://docs.godotengine.org/en/3.0/tutorials/physics/physics_introduction.html

Static body and area should not report contact

**Comment by YeldhamDev** [comment]
@Ranoller Well, then the bug is that is detecting static bodies when Monitorable is enabled then?

**Comment by Ranoller** [comment]
If you trust documentation... Yes

**Comment by mhilbrunner** [comment]
``Monitorable`` seems wonky in other cases too: https://github.com/godotengine/godot/issues/17330

**Comment by rilpires** [comment]
Please do not fix this "bug" !!!

So, story is: I was migrating my (very large) project from 2.x to 3.x, it was hard, OK, and now I think everything is fine, except that I used a RigidBody2D as a "collider detector" and it shouldn't physically collide with other objects (rigid & statics), so it had "trigger" enabled. I had this problem when migrated because in 3.0 there isn't a "trigger" boolean in CollisionShape2D's. Area2D wouldn't do the job because Area2D can't detect static bodies (I had thought so until now), unless Monitorable is enabled.

tldr: I was looking how to detect static bodies with a non-physical colliders and ENABLING MONITORABLE in Area2D is the only way. 

It shouldn't even be called a bug since area2D isn't monitorable anyway(except by others Area2D, I know, but it can be handled with collision_mask bits).

**Comment by rilpires** [comment]
like, I found a solution to my problem in a "bug" topic ... cool thanks btw

**Comment by akien-mga** [comment]
See also test project from duplicate issue: https://github.com/godotengine/godot/files/2217245/StaticBodyAreaBug.tar.gz

**Comment by fossegutten** [comment]
Why is this and the referenced issue #20364 closed? Is it intended that area2D shouldn't detect static bodies and tilemaps?
I just encountered this issue in Godot 3.1.1 and its bugging me. Performance drops by over 50% if I enable monitorable

**Comment by YeldhamDev** [comment]
@fossegutten This issue is currently open, the other one was closed for being a duplicate of this one.

**Comment by alastair-JL** [comment]
So.... what is the best way of detecting static bodies? 
Like, let say I want a character to jump BEFORE running off a ledged, then I want some kind of detector running in front of them that says when there isn't any ground. I don't want the detector to have any physics in it (hence shouldn't be a body), I want it to only be a zone of detection.

This seems like something that "area2d" should do, or at least, have the ability to do. 
If not, then, what type of object should have that behavior.

I'm all for fixing some other thing that is a bug, but ... if you do I'm not sure how to implement the thing I'm imagining.

**Comment by AtomicFirefox** [comment]
> 
> 
> http://docs.godotengine.org/en/3.0/tutorials/physics/physics_introduction.html
> 
> Static body and area should not report contact

Then how would you detect StaticBody2Ds ?

**Comment by wingedadventurer** [comment]
What's the status on this issue? I'm still experiencing it in 3.2.2 stable.

**Comment by char0xff** [comment]
Experiencing this in 3.2.3 stable.

**Comment by char0xff** [comment]
> http://docs.godotengine.org/en/3.0/tutorials/physics/physics_introduction.html
> 
> Static body and area should not report contact

In my case I was trying to detect tiles from a TileMap.
The documentation says:
_body_entered(body: Node)
Emitted when a physics body enters.
The body argument can either be a PhysicsBody2D or a TileMap instance (while TileMaps are not physics body themselves, they register their tiles with collision shapes as a virtual physics body)._
But it doesn't work unless you switch "monitorable" on, on the detector, which makes no sense.

**Comment by madmiraal** [comment]
The problem is with the way the physics engine is designed. To make things efficient, `StaticBodies` are not checked if they're colliding with anything, because they don't move. Instead it relies on the other `CollisionObject`s moving and detecting collisions with it. Similarly, to make things efficient, `Area`s that are not `monitorable` are not checked if they're colliding with anything. The problem arises when moving `Area`s that are not `monitoring` collide with `StaticBodies`. Neither is checking for collisions, so nothing is detected.

**Comment by YeldhamDev** [comment]
@madadam So it's a quirk, rather than a bug? If so, it should be documented.

**Comment by madmiraal** [comment]
No, it's definitely a bug. It's just not an easy one to fix, because of the way the physics engine was designed.

**Comment by char0xff** [comment]
> @madadam So it's a quirk, rather than a bug? If so, it should be documented.

It is a bug, please see https://github.com/godotengine/godot/issues/17238#issuecomment-703217265

**Comment by char0xff** [comment]
>  The problem arises when moving `Area`s that are not `monitoring` collide with `StaticBodies`. Neither is checking for collisions, so nothing is detected.

No. The problem arises even when `Area`s that **are** `monitoring` collide with `StaticBodies`, or at least with tiles, in my case. Nothing is detected until your `Area`s gets switched to `monitorable`, which doesn't make sense, and that's the bug, for what I'm able to understand.

**Comment by madmiraal** [comment]
@char0xff Thanks. I've corrected the typo.

**Comment by Avarobin** [comment]
Why not just remove the option? It rarely works correctly ^^

**Comment by pouleyKetchoupp** [comment]
As a note, the same things happens in Godot Physics 3D, but not when using Bullet (see #50050).

**Comment by barbaros83** [comment]
lol, i was going crazy thinking i did something wrong, turns out its a long standing bug.
is this going to be fixed in 3.x ?
1 of the comments suggests its not a bug, but it definitely is, because the dev should decide what the node should do, not asuming anything.
in my game i have a building that i can place, wich is a staticbody, and the enemy characters are kinematicbody's, with an area for detection, it makes sense the way its set up, and the way you would asume it works, if detection is not desired there are many ways to handle that, put it on a diffrent layer, or disable the body, filter the bodies that enter the area.

**Comment by pouleyKetchoupp** [comment]
@barbaros83 I can confirm this is considered a bug. When it's fixed, it will very likely be on both 4.x and 3.x.

**Comment by lawnjelly** [comment]
Just looking at fixing this now. Here is a minimum reprod project (there was another earlier but it didn't work well for me.

Just run it using either Godot physics or Bullet, it prints the "Area entered" message with Bullet but not with Godot physics.

[Lawn_StaticAreaCollision.zip](https://github.com/godotengine/godot/files/7656040/Lawn_StaticAreaCollision.zip)

**Comment by lawnjelly** [comment]
Incidentally to give more info, the reason this behaviour occurred is that the old octree (and hence the BVH up until #55640) is split into "pairable" and "non-pairable" groups.

### Pairable
* Moving bodies
* monitorable Areas

### Non-Pairable
* Areas
* Statics

Pairable can pair (detect collisions) with pairable
Pairable can pair with non-pairable
Non-pairable cannot pair with non-pairable

Having non-pairable not collide with non-pairable means that if you move one (say an Area) you can omit a large amount of collision checks. The moment you decide you *do* want areas to collide with statics, then you lose out on this performance boost completely.

Bullet presumably doesn't suffer from this because it doesn't make the distinction between pairable and non-pairable (or not in the same way that gains efficiency, I haven't examined the bullet code).

So it comes down to this, currently we need to make a decision, and from what I'm hearing here it sounds like users want everything to collide against everything.

In #55640 I've moved away from pairable and non-pairable, instead using a definable number of trees (up to 32, but only 2 are currently used). For each object added you can then decide which tree it can collide against. I'm still using 2 trees for now for physics, to make the distinction as it may reduce housekeeping, at a cost of having to do collision against 2 trees.

But I have for now set it up so that all objects added collide against both trees. This makes the decision to solve the "bug" at the cost of this efficiency. Although the increased efficiency would only have occurred before if you had moving areas. For statics there is no loss to making them collide with everything, because, by definition they do not move, and except for placing them initially, all the collide tests will be `against them` rather than caused `by them`, which is essentially the same as the current status quo.

We also have the additional ability to specify per-object tree collision behaviour explicitly, which could give much finer control over this kind of thing efficiency wise. Although it is not yet exposed to the user, we have to discuss whether or not we want to do this, as it will probably not be available in bullet.

**Comment by tinmanjuggernaut** [comment]
> So it comes down to this, currently we need to make a decision, and from what I'm hearing here it sounds like users want everything to collide against everything.

The GD4 Area3D documentation says: `3D area that detects CollisionObject3D nodes overlapping, entering, or exiting.` 

So yes, that is what we expect (and how it works in GD3/Bullet). If any PhysicsBody with an enabled CollisionShape and matching layer/mask enters an Area, we should get body entered signals. If any PhysicsBodies collide with other PhysicsBodies with matching parameters, they should collide and send signals.

We also understand that we should be using StaticBodies for immovable objects, RigidBodies for physics movable objects, Kinematic/Character bodies for user movable object, and Areas for detection. Us users don't care about or understand pairable or not pairable or the BVH or the other magic you guys are doing under the hood. We just expect the code to work as advertised in the documentation and tutorials.

**Comment by lawnjelly** [comment]
Saw a few people had been bumping this and similar bugs. Just to review some options for fixing this (as I remember, it's been a while so my memory may be wonky) for the benefit of new physics contributors such as @rburing :

Imo there are two good options which solve the bug (and don't necessarily adversely affect performance in the situation where this mode is not needed). Both involve the need for an extra flag on `Areas`, whether or not the `Area` should collide test against `Statics`. Or alternatively the monitorable state could become more of an enum:

* Normal (as before, Areas in the static tree)
* Collide all (with statics)
* Monitorable (collide with all and report collisions)

The two ways obvious ways of implementing this:
1) Keep the same number of BVH trees (2), but alter the physics code so that when `collide all` is set, the `Area` is added to the dynamic tree instead of the static tree. This would be more expensive than normal mode, but not as expensive as monitable because there is no reporting.
2) Add a third BVH tree (mostly similar to the code I already wrote in #58505 ) and place `collide all` Areas into this third tree instead of the static tree. This has the advantage that we could eliminate `Area` to `Area` collision checks.

Just to emphasize that we made the BVH totally configurable with the templated trees in #55640 , so it is possible to add / remove trees / configure collision masks without changing the actual BVH code at all (and thus requires no specialist knowledge), and this can be done by a physics contributor as in the approach in #58505 .

**Comment by FeldrinH** [comment]
I am not knowledgeable enough about physics engine internals to comment on the technical merits of any choice, but as a user, the idea that 'monitorable' and 'collide all' are two options in the same enum feels weird, since 'monitorable' is about being detectable whereas 'collide all' is about detecting other bodies.

**Comment by Zireael07** [comment]
@FeldrinH that was my first thought upon reading the comment, too

**Comment by AdriaandeJongh** [comment]
Related to https://github.com/godotengine/godot/issues/90077

**Comment by lekoder** [comment]
It might be wise to rethink the approach to bugs like this

I just got hit by this 6-year-old problem that seems to still affect all the stable branches. Perhaps problems like this should be documented in the docs under a "known issues" section or something similar once an issue is open for several months?

My use case was: `Area2D` moving with an NPC ship, used to detect potential collision form the burnfinding to consider.

**Comment by AdriaandeJongh** [comment]
The title of this issue says '3D issue only' but that is incorrect: I'm experiencing this issue in my 2D project in Godot 4.2.2.

**Comment by akien-mga** [comment]
> The title of this issue says '3D issue only' but that is incorrect: I'm experiencing this issue in my 2D project in Godot 4.2.2.

"(3D issue only when using GodotPhysics)" means that the issue affects 3D (it's a "3D issue") _only_ when using GodotPhysics (i.e. not Bullet, in 3.x, at the time where this title was edited).

I'll reword for clarity.

**Comment by Very-Blank** [comment]
Seems like _set_static() is the issue.
```
void GodotArea3D::set_monitorable(bool p_monitorable) {
	if (monitorable == p_monitorable) {
		return;
	}

	monitorable = p_monitorable;
	_set_static(false); // was -> _set_static(!monitorable)
	_shapes_changed();
}
```
Just setting it to false seems to work at least with a basic test scene. I don't now if the "issue" is indented but the docs only say this about "monitorable":

> If true, other monitoring areas can detect this area.

**Comment by lock3r** [comment]
Just stumbled upon this. Turns out the documentation is misleading and if you don't set monitorable to true the Area3D doesn't detect anything (not Static Bodies, not other areas - nothing). I think that, since this is a 6 years old behavior, it should at least be documented.

**Comment by aleksa2808** [comment]
> Just stumbled upon this. Turns out the documentation is misleading and if you don't set monitorable to true the Area3D doesn't detect static bodies. I think that, since this is a 7 years old behavior, it should at least be documented.

Agreed, a note in the documentation would be very helpful, assuming of course that moving Area2Ds are recommended, not just a common practice. Even when I understand what needs to be done it still feels off for the Area2D to be monitorable when using it just to detect other bodies.
